snapshot/snapshot.sh

```diff
@@ -0,0 +1,353 @@
+    if [ "$status" != "Running" ]; then 
+        echo "Error starting target pod"
+        exit 1
+    fi
+
+    ip=$(kubectl get pod -n $KONTAIN_NAMESPACE $TARGET_POD_NAME -o jsonpath='{.status.podIP}')
+    port=$(kubectl get pod -n $KONTAIN_NAMESPACE $TARGET_POD_NAME -o jsonpath='{.spec.containers[0].ports[0].containerPort}')
+    port=${port:=80}
+
+    declare -A extra_kontain_envs
+    extra_kontain_envs[IP]=$ip
+    extra_kontain_envs[PORT]=$port
+
+    add_envs $MAKER_POD_YAML extra_kontain_envs
+
+    # assign pod to specific node
+    set_node_selector $MAKER_POD_YAML
+
+    kubectl apply -f $MAKER_POD_ENTRYPOINT_YAML >& /dev/null
+    kubectl apply -f $MAKER_POD_YAML >& /dev/null
+
+    # wait for manager pod to be ready, i.e untill snapshot is made 
+    kubectl wait --for=condition=Ready pod -n $KONTAIN_NAMESPACE $MAKER_POD_NAME >& /dev/null
+    status=$(kubectl get pod -n $KONTAIN_NAMESPACE $MAKER_POD_NAME -o jsonpath="{.status.phase}")
+    if [ "$status" != "Running" ]; then 
+        echo "Error starting maker pod"
+        exit 1
+    fi
+
+    echo "done"
+}
+
+function make_image() {
+
+    echo -n "Generating a new snapshot-based image....."
+
+    mkdir -p $IMAGE_DIR
+    kubectl cp $KONTAIN_NAMESPACE/$MAKER_POD_NAME:$SNAPDIR_CONTAINER $IMAGE_DIR >& /dev/null
+
+    # Create a new container image to run the snapshot
+    # This needs to handle differing snapshot filenanes kmsnap vs kmsnap.XXX.NNN
+    # Build dockerfile
+    # Build container image
+    # Push the container to a remote repository
+    SNAPFILE=$(basename $IMAGE_DIR/kmsnap*)
+
+    cat <<EOF >$SNAP_DOCKER_FILE
+FROM $IMAGE_NAME
+COPY $SNAPFILE /kmsnap
+CMD ["/kmsnap"]
+EOF
+
+    
+    SNAP_IMAGE_NAME=$KONTAIN_SNAPSHOT_REGISTRY/$(basename "${IMAGE_NAME%%:*}")_snap
+
+    chown $(id -u):$(id -g) $IMAGE_DIR/$SNAPFILE
+    chmod 755 $IMAGE_DIR/$SNAPFILE
+
+    result=$(docker build -t $SNAP_IMAGE_NAME -f $SNAP_DOCKER_FILE $IMAGE_DIR 2>&1 > /dev/null)
+    if [ $? != 0 ]; then
+        echo "Failed to build snapshot image"
+        echo "docker: $result"
+        exit 1
+    fi
+    echo "done"
+}
+
+function upload_image() {
+    echo -n "Uploading image to dockerhub....."
+
+    result=$(docker push $SNAP_IMAGE_NAME 2>&1 > /dev/null)
+    if [ $? != 0 ]; then
+        echo "Failed to push image to the repository"
+        echo "Make sure container registry $KONTAIN_SNAPSHOT_REGISTRY exists and you are logged into it with necessary permissions"
+        echo "docker: $result"
+        exit 1
+    fi
+    echo "Image $SNAP_IMAGE_NAME has been uploaded"
+
+}
+function update_pods() {
+
+    echo -n "Updating deployment with new image....."
+
+    kubectl set image deployment.apps/$deployment_name $CONTAINER_NAME=$SNAP_IMAGE_NAME >& /dev/null
+    
+    echo "done"
+}
+
+function main() {
+
+    load_tools
+
+    kubectl create namespace $KONTAIN_NAMESPACE >& /dev/null
```

Comment:
If we send the error messages to /dev/null, how will we know went wrong, so that we can fix the problem?